src/libpanic_unwind/gcc.rs

```diff
@@ -175,6 +172,23 @@ unsafe extern "C" fn rust_eh_personality(version: c_int,
     }
 }
 
+// The personality routine for most of our targets, except ARM, which has a slightly different ABI
+// (however, iOS goes here as it uses SjLj unwinding).  Also, the 64-bit Windows implementation
+// uses a different personality (below) but eventually also calls rust_eh_personality_impl.
+#[cfg(all(any(target_os = "ios", target_os = "netbsd", not(target_arch = "arm")),
+          not(all(windows, target_arch = "x86_64", target_env = "gnu"))))]
```

Comment:
Would it be possible perhaps while you're here to use `cfg_if!` to clean up this `#[cfg]` clause?

Reply:
> The [OP description](https://github.com/rust-lang/rust/pull/65646#issue-330187431) was also very descriptive, thanks! Can you also explain in a bit more detail what's going on with MinGW? All the changes here I was able to match up with the description but I want to just double check my interpretation of MinGW.

For 32-bit MinGW there is no change (apart from the handling of foreign exceptions in the personality function). For 64-bit MinGW we want to use the same exception mechanism as C++ code (libunwind) instead of rolling our own thing. Conveniently, libunwind already has a translation layer (`_GCC_specific_handler`) which wraps a libunwind-style personality function and hides all the ugly SEH details. This enables:
- us to run Rust destructors when a C++ exception unwinds, but not for arbitrary SEH exceptions (such as the one used by `longjmp`).
- C++ destructors to run when a Rust panic unwinds, since `_GCC_specific_handler` only passes exceptions with code `STATUS_GCC_THROW` and `STATUS_GCC_UNWIND` to the libunwind-style personality function.

> Also you mention `longjmp` on MSVC, where I believe we intentionally don't want to run Rust destructors on `longjmp` (it doesn't anywhere else so I figured it shouldn't on MSVC). I didn't really see any changes which would lead to that though, so I'm curious is this something that's intented as a result of this PR?

Actually `longjmp` does run destructors (both C++ and Rust) because it uses a SEH exception to unwind. This was already the case before this PR. However `_CxxFrameHandler3` specifically ignores `longjmp` exceptions for `catch` blocks.

This behavior is documented in MSVC, and is allowed by the C++ standard since it is UB to unwind across frames with destructors.